pkg/system/system.go

```diff
@@ -470,35 +468,73 @@ func buildNextAPISocket(cur string) (string, error) {
-// copyNydusdBinary copies a file from sourcePath to destinationPath,
-// ensuring parent directories exist and setting 0755 permissions.
-// It overwrites the destination file if it already exists.
-func copyNydusdBinary(sourcePath, destinationPath string) error {
-	fileMode := os.FileMode(0755)
+// upgradeNydusdWithSymlink atomically upgrades the nydusd binary using a symbolic link strategy.
+// It copies the new binary to a permanent, versioned directory and then atomically updates a symlink
+// to point to the new version. This allows for zero-downtime upgrades and avoids "text file busy" errors.
+func upgradeNydusdWithSymlink(newBinaryPath, symlinkPath string) error {
+	// Determine the directory for storing all actual binary files.
+	// This is typically a subdirectory within the same directory as the symlink.
+	symlinkDir := filepath.Dir(symlinkPath)
+	versionsDir := filepath.Join(symlinkDir, "nydusd-versions")
+	if err := os.MkdirAll(versionsDir, 0755); err != nil {
+		return fmt.Errorf("failed to create versions directory %s: %w", versionsDir, err)
+	}
+
+	// Generate a unique name for the new binary file
+	// and determine its final destination path inside the versions directory.
+	versionedFileName := fmt.Sprintf("nydusd-%d", time.Now().UnixNano())
+	versionedBinaryPath := filepath.Join(versionsDir, versionedFileName)
 
-	destDir := filepath.Dir(destinationPath)
-	if err := os.MkdirAll(destDir, fileMode); err != nil {
-		return fmt.Errorf("failed to create destination directory %s: %w", destDir, err)
+	// Copy the new binary file to its destination in the versions directory.
+	if err := copyFile(newBinaryPath, versionedBinaryPath); err != nil {
+		return fmt.Errorf("failed to copy new binary to versions directory: %w", err)
 	}
+	log.L.Infof("Successfully copied new binary to %s", versionedBinaryPath)
+
+	// Atomically update the symbolic link.
+	// To do this safely, we first create a temporary symlink and then use an atomic rename
+	// operation to replace the old symlink with the new one.
+	tempSymlinkPath := symlinkPath + ".tmp-" + xid.New().String()
+	if err := os.Symlink(versionedBinaryPath, tempSymlinkPath); err != nil {
+		return fmt.Errorf("failed to create temporary symlink %s: %w", tempSymlinkPath, err)
+	}
+
+	// os.Rename provides an atomic replacement of the symlink.
+	if err := os.Rename(tempSymlinkPath, symlinkPath); err != nil {
+		// Clean up the temporary symlink if the rename fails.
+		os.Remove(tempSymlinkPath)
+		return fmt.Errorf("failed to atomically update symlink %s: %w", symlinkPath, err)
+	}
+
+	log.L.Infof("Successfully updated symlink %s to point to %s", symlinkPath, versionedBinaryPath)
+	return nil
+}
+
+// copyFile is a utility function that copies a file from a source path to a destination path.
+// It ensures the destination directory exists and sets the destination file's permissions to 0755.
+// If the destination file already exists, it will be overwritten.
+func copyFile(src, dst string) error {
```

Comment:
How about trying `os.Link` to create a hard link first, then fall back to `io.Copy` if it fails (e.g., due to cross-device limitations)?